[qt] move QSettings to bitcoin_rw.conf where possible

Sjors opened the pull request:
When QT launches, move the following QSettings to `bitcoin_rw.conf` (added in #11082):

-`lang` (language)
-`dbcache` (nDatabaseCache)
-`par` (nThreadsScriptVerif)
-`prune` (bPrune,bPruneSize)
-`spendzeroconfchange` (bSpendZeroConfChange)
-`upnp` (fUseUPnP)
-`listen` (fListen)
-`proxy` (fUseProxy)
-`onion` (fUseSeparateProxyTor)

~This is a step towards being able to configure `prune` from the GUI (#6461). ~ (already done )

There's potentially a safety benefit in storing  `listen`, `proxy` and `onion` in `bitcoin_rw.conf` in case the user ever decides to run `bitcoind`.

luke-jr committed:
Add new bitcoin_rw.conf file that is used for settings modified by this software itself

laanwj commented:
Please, never write to `bitcoin.conf`. It is exceptionally bad form for a daemon to write to its configuration file. If you want to do this, I'd suggest adding an additional file (in the datadir) for read/write settings (as @luke-jr did AFAIK).

Sjors commented:
@laanwj two files is indeed what @luke-jr did, but I don't see how that changes anything. It's critical that `bitcoind` uses this second file (for settings like `prune`) and treats it the same as `bitcoin.conf`. So by the same reasoning it shouldn't have that second file.

I thought of a completely different approach to handle the `prune` case, so we might be able to kick this shared settings can down the road. Will open a PR for later.

laanwj commented:
> @laanwj two files is indeed what @luke-jr did, but I don't see how that changes anything. 

One thing is permissions. The data directory is guaranteed writable. The configuration file might be `-conf=/etc/bitcoind.conf`.

> Also note that only QT needs this ability, so I could refactor it such that bitcoind daemon only has read-only access.

Unlike QSettings the second file would be used by both the GUI and the daemon. 

And it might be useful to edit settings in bitcoind at real-time as well. The GUI shouldn't really be special there, that gets in the way of the whole idea of unifying the settings mechanism.

Sjors commented:
If `bitcoin.conf` is not writeable, because the user configured a custom read-only directory, the UI elements that change settings this way could be disabled.

> And it might be useful to edit settings in bitcoind at real-time as well.

Indeed.

I thought having two separate files leads to a more complicated implementation than just having one, but I didn't remove that much from the original code so far. It helps that settings can be overridden by repeating them in the new file.

So in that case it might more sense to get #11082 merged, in which case this PR just adds QSettings migration code on top of that.

Sjors commented:
I pushed a new version that uses `bitcoin_rw.conf`. I could probably add some helper functions to make the change more compact.

jonasschnelli commented:
I'm not sure if moving the QT settings from pure QT format in the QT chosen file to the bitcoind compatible `bitcoin_rw.conf` format reduces the complexity. Configuration file complexity can be harmful (-datadir, eventually once config files per network, includes, etc.).

I also think we could expose prune to the QT layer without sharing the configuration value with bitcoind.

Users using bitcoind & QT with the same settings and datadir do probably know the ramification of GUI only settings. A warning when enabling prune in QT (once implemented) that it will not automatically be enabled in daemon mode seems sufficient to me.

Sjors commented:
@jonasschnelli in addition to a warning, we can even make `bitcoind` actually handle this elegantly, see my suggestion [here](https://github.com/bitcoin/bitcoin/issues/6461#issuecomment-377485323).

For settings other than prune there might be similar workarounds on a case by case basis (other than warning).

Sjors commented:
This and #11082 need a likely non-trivial rebase after #11862. I'm going to try the other approach first though.

Sjors commented:
#11862 has been merged. I'll rebase this once #11082 is rebased.

ryanofsky commented:
I like this approach. I think it'd be a simplification to stop using the windows registry and other config backends in `bitcoin-qt`. And it seems like it would be less surprising and error prone if `bitcoin-qt`  and `bitcoind` used the same settings, instead of slightly different but overlapping sources of settings like they do now, so I don't get the objection in https://github.com/bitcoin/bitcoin/pull/12833#issuecomment-378581605.

One thing I don't understand about the implementation though is why the "Active command-line options that override above options" display is removed. Will command line settings not override config file settings anymore with this change? I think it would be bad if that were the case.

Sjors commented:
@ryanofsky wrote:
> Will command line settings not override config file settings anymore with this change?
They will. The phrase "active command-line options" actually referred to both `bitcoin.conf` and command line options, either of which can override QT settings, as well as each other.

Afaik QT can't tell the difference between settings that came from bitcoin.conf and those that came from a command line option, which is why I removed that text. It's also pretty difficult in practice to launch QT with command line options, involving editing desktop shortcuts etc.

luke-jr committed:
util: Support prepending configs in ReadConfigStream

luke-jr committed:
util: SelectBaseParams in ReadConfigFiles, before getting final datadir

luke-jr committed:
util: ReadConfigFiles flag to make everything network-specific

Sjors commented:
Rebased and added pruning to the list of imported settings.

I changed the meaning of _disabling_ prune in the GUI to mean _stop further pruning_, instead _undo previous pruning_. It just sets `prune=1`, which avoids forcing the user to re-download the chain. We could later add a button to re-download the whole chain.

practicalswift commented on src/qt/optionsmodel.cpp:
```diff
@@ -420,32 +397,55 @@ bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, in
+              // size in settings.
+              nPruneSizeMB = gArgs.GetArg("-prune", (qint64)nDefaultDbCache);
+              settings.setValue("nPruneSize", nPruneSizeMB / 1000);
+              // TODO: check if pruning actually occured, otherwise prune=0 is safe too
```
"Occurred" :-)

practicalswift commented on src/qt/optionsmodel.cpp:
```diff
@@ -420,32 +397,55 @@ bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, in
+            if (value.toBool()) {
+              // When user enables pruning, store the prune size:
+              nPruneSizeMB = settings.value("nPruneSize").toInt() * 1000;
+              gArgs.ModifyRWConfigFile("prune",std::to_string(nPruneSizeMB));
```
Nit: Missing whitespace after `,` :-)

practicalswift commented on src/util/system.cpp:
```diff
@@ -979,6 +1028,217 @@ std::string ArgsManager::GetChainName() const
+            return false;
+        }
+        result.clear();
+        result.push_back(char(current_char));
```
Nit: I believe `(char)current_char` is the preferred form for casting the fundamentals types in the project :-)

Sjors replied:
That's upstream: #11082

practicalswift commented on src/util/system.cpp:
```diff
@@ -979,6 +1028,217 @@ std::string ArgsManager::GetChainName() const
+            if (current_char == std::char_traits<char>::eof()) {
+                break;
+            }
+            result.push_back(char(current_char));
```
Same here :-)

practicalswift commented on src/qt/optionsmodel.cpp:
```diff
@@ -243,6 +171,7 @@ static const QString GetDefaultProxyAddress()
 // read QSettings values and return them
 QVariant OptionsModel::data(const QModelIndex & index, int role) const
 {
+    int64_t prune = -1;
```
The scope of `prune` can be limited to the scope of `case Prune` and `case PruneSize` below :-)

practicalswift commented on src/qt/optionsmodel.cpp:
```diff
@@ -340,8 +283,16 @@ bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, in
         case ProxyUse:
-            if (settings.value("fUseProxy") != value) {
-                settings.setValue("fUseProxy", value.toBool());
+            if (value.toBool() ^ gArgs.GetArg("-proxy", "") != "") {
```
Add parentheses to make precedence obvious :-)

practicalswift commented on src/qt/optionsmodel.cpp:
```diff
@@ -359,15 +314,27 @@ bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, in
         case ProxyUseTor:
-            if (settings.value("fUseSeparateProxyTor") != value) {
-                settings.setValue("fUseSeparateProxyTor", value.toBool());
+            if (value.toBool() ^ gArgs.GetArg("-onion", "") != "") {
```
Same here :-)

practicalswift commented on src/qt/optionsmodel.cpp:
```diff
@@ -420,32 +397,55 @@ bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, in
         case DatabaseCache:
-            if (settings.value("nDatabaseCache") != value) {
-                settings.setValue("nDatabaseCache", value);
+            if ((int64_t)gArgs.GetArg("-dbcache", (qint64)nDefaultDbCache) != value) {
```
Is this cast to `int64_t` needed?

practicalswift commented on src/qt/optionsmodel.cpp:
```diff
@@ -420,32 +397,55 @@ bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, in
         case ThreadsScriptVerif:
-            if (settings.value("nThreadsScriptVerif") != value) {
-                settings.setValue("nThreadsScriptVerif", value);
+            if ((int64_t)gArgs.GetArg("-par", DEFAULT_SCRIPTCHECK_THREADS) != value) {
```
Is this cast to `int64_t` needed?

Sjors commented:
I'll rebase and fix code issues after the next upstream PR rebase.

ryanofsky commented:
Interesting! Apparently I commented on this PR and then forgot it existed, because I reimplemented the same thing six months later in #15936

Sjors committed:
Add havePruned() to Node interface

Sjors committed:
[gui] settings: remove command line override label

addOverriddenOption becomes an empty function to keep the diff simple, due to the many if statements without brackets. It will be removed in the next commit.

Sjors committed:
[qt] move QSettings to bitcoin_rw.conf where possible

When QT launches, move the following QSettings to
bitcoin_rw.conf:

-lang (language)
-dbcache (nDatabaseCache)
-par (nThreadsScriptVerif)
-prune (bPrune,nPruneSize)
-spendzeroconfchange (bSpendZeroConfChange)
-upnp (fUseUPnP)
-listen (fListen)
-proxy (fUseProxy)
-onion (fUseSeparateProxyTor)

Disabling pruning in the GUI now sets prune=1 if the chain is pruned, otherwise prune=0.

Sjors committed:
[doc] clarify control flow in AppTests

Sjors committed:
[wallet] move constants to walletconstants.h

jonasschnelli commented:
Closing for now since this is dependant on a dormant PR #11082

Sjors commented:
I'll nag to reopen this once 11082 is rebased.

Sjors commented:
@jonasschnelli 11082 is rebased, can you re-open this?

Rspigler commented:
@Sjors are you still working on this/asking this to be opened, with #15935 being merged and #15936 being discussed?
